spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java

```diff
@@ -665,16 +670,34 @@ protected void asyncFailure(Object request, @Nullable Acknowledgment acknowledgm
 			Throwable t, Message<?> source) {
 
 		try {
-			handleException(request, acknowledgment, consumer, source,
-					new ListenerExecutionFailedException(createMessagingErrorMessage(
-							"Async Fail", source.getPayload()), t));
+			if (t instanceof CompletionException) {
+				// For CompletableFuture Object
+				handleException(request, acknowledgment, consumer, source,
+								new ListenerExecutionFailedException(createMessagingErrorMessage(
+										"Async Fail", source.getPayload()), t.getCause()));
+			}
+			else {
+				// For Mono Object.
+				handleException(request, acknowledgment, consumer, source,
+								new ListenerExecutionFailedException(createMessagingErrorMessage(
+										"Async Fail", source.getPayload()), t));
```

Comment:
When `CompletableFuture<?>` fails to complete due to an unexpected errors, it throws a `CompletionException`.  By the way, when `Mono<?>` fails to complete due to an unexpected errors, it throws the root cause exception. 

It makes difference when `KafkaMessageListenerContainer` executes retry.
For example:

`KafkaMessageListenerContainer.handleAsyncFailure()`  -> ... -> `FailedRecordTracker.recovered()` -> `FailedRecordTracker.attemptRecovery()` -> `this.recoverer.accept(...)` -> `DeadLetterPublishingRecoverer.accept(...)` -> `this.destinationResolver.apply(...)` -> `DeadLetterPublishingRecoverereFactory.destinationResolver(...)` -> `this.destinationTopicResolver.resolveDestinationTopic(...)`.

In this case, `destinationTopicHolder.getSourceDestination().shouldRetryOn(attempt, maybeUnwrapException(e))` returns `false` when the exception is a `CompletionException`.
This is because `BinarayExceptionClassifier.classify(...)` and its superclass `SubclassClassifier` return false for `CompletetionException`. 


```java
	@Override
	public DestinationTopic resolveDestinationTopic(String mainListenerId, String topic, Integer attempt, Exception e,
													long originalTimestamp) {
		DestinationTopicHolder destinationTopicHolder = getDestinationHolderFor(mainListenerId, topic);
		return destinationTopicHolder.getSourceDestination().isDltTopic()
				? handleDltProcessingFailure(destinationTopicHolder, e)
				: destinationTopicHolder.getSourceDestination().shouldRetryOn(attempt, maybeUnwrapException(e)) // This is false with CompletionException.
						&& isNotFatalException(e)
						&& !isPastTimout(originalTimestamp, destinationTopicHolder)
					? resolveRetryDestination(mainListenerId, destinationTopicHolder, e)
					: getDltOrNoOpsDestination(mainListenerId, topic, e); // thus, CompletionException gets the DLT Topics as next destination.
	}
```
Therefore, when `CompletableFuture` throws a `CompletionException`, `resolveDestinationTopic()` always return DLT topics because `destinationTopicHolder.getSourceDestination().shouldRetryOn(attempt, maybeUnwrapException(e))` is false. 


In short, 
1. `CompletableFuture` and `Mono` throw different exception when they fail to complete.
2. It makes difference when async retry. (`CompletableFuture` -> Always DLT,  `Mono` -> retryable topic and DLT).
3. Also, I think `CompletionException` is not root cause and the classifier seems to not consider root cause when they classify.

Thus, we should unwrap the `CompletionException` once in `asyncFailure()`.